### DISCUSSION TITLE
Help: How to use App.vue and wait for store to be ready through the app

### DISCUSSION THREAD

**Discussion by nosizejosh** [nuxt/nuxt#16257] [Questions] [answered]
### Environment

- Operating System: `Windows_NT`
- Node Version:     `v16.17.1`
- Nuxt Version:     `3.0.0`
- Nitro Version:    `1.0.0`
- Package Manager:  `yarn@1.22.19`
- Builder:          `vite`
- User Config:      `runtimeConfig`, `postcss`, `css`, `modules`
- Runtime Modules:  `@formkit/nuxt@1.0.0-beta.9-f64c966`, `@vueuse/nuxt@9.6.0`, `@pinia/nuxt@0.4.5`, `@nuxtjs/supabase@0.3.0`, `@nuxtjs/tailwindcss@6.1.3`
- Build Modules:    `-`

### Reproduction

![image](https://user-i
![edit](https://user-images.githubusercontent.com/392227/204125685-0cd2e9cc-e7cb-4040-8c04-dbc83d387c05.png)
mages.githubusercontent.com/392227/204125472-48a9e7d2-8c8e-43d6-a2ae-1ded2e9ed844.png)

### Describe the bug

Hello All, and @danielroe  @aldarund from reading your guides and documentation, app.vue is the entry point of the nuxt 3 app and so if I stored some profile information in a store and loaded it in app.vue, then it should be readily accessible from other pages right? I am facing a race condition where some pages are loaded before app.vue or their call to this store variable happens before app.vue and so their value is null at that point. 

1. How do I makes sure this variable which is from an external server is ready for all pages and not just app.vue
2. If app.vue is the entry point, why do some pages run before it, leading to the variable being null at that point?

this is how my app.vue looks like 
![image](https://user-images.githubusercontent.com/392227/204125304-1bf4683f-1c6a-425d-95da-3201072b43f9.png)
 and a console log 
![edit](https://user-images.githubusercontent.com/392227/204125694-3630ad3a-5c63-4bb3-8a7e-ce63788fdcd9.png)

that shows clearly store variable userWithProfile is called in ticketTable where it returns null before the console.log in  app.vue where it is properly set and returns a value. 

3. How can I ensure the variables in app are properly set before the rest of the app runs as they are required for proper function of the app?

Please help

### Additional context
Noticed that if another route is visited and then returned to this tickets page, then the userWithProfile is now properly set on the page, but it was set from the start on app.vue.

**Comment by TheAlexLichter** [comment]
A few things:

1. Please don't tag maintainers in the post to ping them 😋 We read through all the issues and discussion!
2. If you `await` in `onMounted` (or in lifecycle hooks in general), child components (and ones further down) won't wait. You need to do that directly in the `<script setup>` part.

**Reply by nosizejosh** [reply]
<script await setup> worked. Thank you.
Sorry for tagging :)

**Reply by nosizejosh** [reply]
I'm I using app.vue correctly though?
By forcing await in the setup, i'm I not going to run into performance issues later on?
For example, when the entire app has to await for all async calls to be ready? even those not currently in view or active page, since app.vue runs for all pages?
Is there a performant way to achieve what I am looking for? -a way to initialize some variables and make sure they are ready and available through the app

**Reply by nosizejosh** [reply]
spoke too soon, refreshing the page makes store unavailable and I think that has to do with store persistence right? 
store variable is available when you route to page but unavailable when page is refreshed

**Reply by nosizejosh** [reply]
tried solving suspected persistence issues with https://vueuse.org/core/useStorage/ but didn't work. this https://prazdevs.github.io/pinia-plugin-persistedstate/frameworks/nuxt-3.html worked. Thanks!

**Reply by TheAlexLichter** [reply]
@nosizejosh Feel free to mark the reply as "answer" then 😋

>  By forcing await in the setup, i'm I not going to run into performance issues later on?

Well, the whole application will wait so everything done in the method you `await` is blocking your app.

**Reply by nosizejosh** [reply]
- Taking out the await in the setup tag and leaving the rest of the code as is works, after installing and persisting the store with https://github.com/prazdevs/pinia-plugin-persistedstate
- So maybe that was the issue from the start?
- Running setup outside of the mounted works too but the loader is client side only so code has to be refactored to take out the loader 
- Not sure which to use, so for now will just run as is without the await in the script setup tag

**Comment by motea927** [comment, marked as answer]
1. Just await in `<script setup>` like this
```ts
<script setup lang="ts">
await setUserUp(loader)
</script>
```
and if you refresh page, it will invoke twice (server-side & client side),
so you need add some logic to ensure variable is existed,
if you use state management(pinia...), you can use `useAsyncData` like below, and return value is **necessary**
```ts
<script setup lang="ts">
await useAsyncData('getProfile', async () => {
  await userStore.getProfile();
  return userStore.id;
});
</script>
```
2.  maybe you can undersatnd more vue lifecycle
- https://medium.com/@brockreece/vue-parent-and-child-lifecycle-hooks-5d6236bd561f

3. reference point 1

**Reply by nosizejosh** [reply]
- Taking out the await in the setup tag and leaving the rest of the code as is works, after installing and persisting the store with https://github.com/prazdevs/pinia-plugin-persistedstate
- So maybe that was the issue from the start?
- Running setup outside of the mounted works too but the loader is client side only so code has to be refactored to take out the loader 
- Not sure which to use, so for now will just run as is without the await in the script setup tag

**Reply by nosizejosh** [reply]
had to rewrite the methods with guards as suggested and also use the persist plugin.
